### ISSUE TITLE
HIGH ENHANCEMENT - In addition to MERGE in Repeats include REPLACE as key_policy

### ISSUE THREAD

**Issue by atl2017** [smap-consulting/smapserver#352]
### Actual Behaviour:
![image](https://user-images.githubusercontent.com/29519901/44216470-a13b6700-a13a-11e8-9337-41709b34d175.png)

### Steps to reproduce the problem:
Following form is used to enable codes from a list using select_multiple and then with a begin repeat moves forward each code a get data with pulldata. The form should be edited as many times as user want. An alert is displayed if former data exists. Former selected data are shown as default (using calculate with once) in order to let unselect or add more by selecting new ones.

### Form Name: ENABLED CODE LIST

An error in record not marked as _bad in subform was found after 4 rounds. FT behaves as expected. 

  | 4A0020 | 4A0040 | 4A0320 | Others | FT (Displayed as selected) | Analysis (Unselected Marked as   _bad)
-- | -- | -- | -- | -- | -- | --
Round 1 | Selected | Unselected | Unselected | Unselected |   |  
Round 2 | Unselected | Selected | Unselected | Unselected | Ok | yes
Round 3 | Selected | Selected | Unselected | Unselected | Ok | yes
Round 4 | Unselected | Unselected | Selected | Unselected | Ok | yes 4A0040 No 4A0020

*You can erased data and reload form if necessary to reproduce results.*

**Comment by atl2017** [comment]
I have upload another form to try to guess what is happenning

 Server: Kontrolid
Organisation: Zarkman
Project: Merge in Repeats
Form Name: REPEAT MERGE YES

The idea of the form is to simulate some checklist that can be modified later. 

Basically form collect an ID, then a select multiple items. Afterwards a subform is used to get data from CSV for each selected item. I have included above twice in order to evaluate with and without using calculation column to let former data be included as dynamic default.

The id (key policy: merged) is used to get former data through a linked_self

![image](https://user-images.githubusercontent.com/29519901/44277545-b1257a80-a210-11e8-999a-7dac7a290072.png)

After 3 rounds using the same Id = 1:
![image](https://user-images.githubusercontent.com/29519901/44277793-6bb57d00-a211-11e8-862a-86ab4ec85266.png)

Analysis table - main:

![image](https://user-images.githubusercontent.com/29519901/44277911-e9798880-a211-11e8-8752-bc5ec9ebe021.png)

Analysis table - x = Analysis table - x2:
![image](https://user-images.githubusercontent.com/29519901/44277956-0b730b00-a212-11e8-8699-7ce554821b0a.png)


Conclusion:
- Behaviour is the same with and without calculation as default
- merge=yes works only when following round is to include an item. I mean move from selected 2 to selected 3
- merge=yes DOES NOT work when following round is to erase(unselect) an item. I mean move from selected 3 to selected 2 still records 3 as selected. FT works fine displayingd only 2 selected. Situation is 
with table.

**Comment by nap2000** [comment]
Great analysis work!  This was really useful.   It seems that the behaviour is currently as intended.   The particular use case I am supporting is the following:

- There is a piece of equipment that receives regular maintenance visits
- Mostly static information on the equipment is stored in the main form
- There is a "begin repeat" that stores data on each visit
- After the visit a supervisor is notified and they get to see and review only what was entered in that last visit.  So not the 10 preceding visits.  They add their approval and confirm actions which result in the data for that last visit being updated.   However none of the records for the preceding 10 visits are affected.

If I read your analysis right you are now requesting support for a use case where only the repeats for the last visit are retained?   I think this can be done.  Perhaps there could be a "replace=yes" parameter.  Or possibly better have a key_policy=merge, key_policy=replace that can be set on the begin repeat.

**Comment by atl2017** [comment]
First, Thanks for detailed explanation of merge = yes. Could you provide an xlsform of your example by Email? I think I could use it.

Second, yes I am requesting support for a use case where only the repeats for the last visit are retained. key_policy=merge, key_policy=replace could be better. I wonder if behaviour is a little bit different from key_policy= merge in main that acts as replace. 

Question: Why, marked as bad is duplicated. Refer to first image signed as "twice"?

**Comment by nap2000** [comment]
I think merge in repeats works the same as merge in the main form.  Its only that there is always one main record and only one main record.

**Comment by atl2017** [comment]
Understood.  You are right!. I just change title of issue from BUG to ENHANCEMENT ans remains open

**Comment by nap2000** [comment]
On begin repeats you can now use:
   key_policy=merge
   key_policy=replace
   key_policy=append

append is the default.

**Comment by atl2017** [comment]
Start testing. key_policy=replace  fails in form above: 

Organisation: Zarkman
Project: Merge in Repeats
Form Name: ENABLED CODE LIST

After second Round it was supposed that one code should replace the other, but both remains:

4A0020 | 4A0040 | 4A0320 | Others | FT (Displayed as selected) | Analysis (Unselected Marked as _bad)
-- | -- | -- | -- | -- | --
Round 1 | Selected | Unselected | Unselected | Unselected |   |  
Round 2 | Unselected | Selected | Unselected | Unselected | Ok | yes

![image](https://user-images.githubusercontent.com/29519901/44310388-b0b8eb00-a39a-11e8-9aec-b695dbe384d3.png)

**Comment by nap2000** [comment]
Ok should work now.  The key policies were not being set if you loaded the form as an xlsForm

**Comment by atl2017** [comment]
Tested Ok. It works!

But I have a different issue while testing: When move from round 2 (3 selected) to round 3 (none selected) selection remains. 

![image](https://user-images.githubusercontent.com/29519901/44339135-e961ce80-a445-11e8-9cfd-741fdb093e57.png)

I Thought it could be related with calculation as dynamic default:

![image](https://user-images.githubusercontent.com/29519901/44339505-385c3380-a447-11e8-987a-c9a44c9f1e6f.png)

But it din't. I have tested with new form: Form Name: REPEAT MERGE YES v3 (cloned from v2) without dynamic default. The results are the same:

![image](https://user-images.githubusercontent.com/29519901/44339635-c2a49780-a447-11e8-9d7c-1c4694553831.png)

So this behaviour are related with key_police:: merge in main table: when a selection question is empty is replaced with former data. **So this policy in main does not permit select then unselect all cases**
